File: src/py/power_analysis_app.py
from dash import Dash, Input, Output, dcc, html
import numpy as np
import plotly.express as px

from .simulate import HEDGES_G, simulate_stroop
from .analysis import (
    stroop_bayes_design,
    stroop_bayes_probability,
    stroop_bayes_t_test,
    stroop_power,
    stroop_t_test,
)


app = Dash(__name__)

app.layout = html.Div([
    html.H2("Stroop simulation and design analysis"),
    html.P(
        "Use the sample controls to simulate reaction times and inspect the group distributions. "
        "Then choose the design criteria to compare frequentist power with the probability of reaching a Bayes-factor threshold."
    ),

    html.H3("Sample characteristics"),
    html.Div([
        html.Div([html.Label("Participants per group"), dcc.Slider(10, 200, 10, value=50, id="n")]),
        html.Div([html.Label("Control mean (ms)"), dcc.Slider(400, 800, 25, value=600, id="mean")]),
        html.Div([html.Label("Pooled SD (ms)"), dcc.Slider(25, 200, 25, value=100, id="sd")]),
        html.Div([html.Label("Hedges' g"), dcc.Slider(0.1, 1.5, 0.01, value=HEDGES_G, id="g")]),
    ], style={"display": "grid", "gridTemplateColumns": "repeat(2, minmax(0, 1fr))", "gap": "24px"}),

    dcc.Graph(id="distribution", style={"width": "100%"}),
    html.P(id="results"),

    html.H3("Design criteria"),
    html.Div([
        html.Div([html.Label("Target probability"), dcc.Slider(0.6, 0.95, 0.05, value=0.8, id="desired-power")]),
        html.Div([html.Label("Frequentist alpha"), dcc.Slider(0.01, 0.10, 0.01, value=0.05, id="alpha")]),
        html.Div([
            html.Label("Bayes factor threshold"),
            dcc.RadioItems([
                {"label": "BF10 ≥ 3", "value": 3},
                {"label": "BF10 ≥ 10", "value": 10},
                {"label": "BF10 ≥ 30", "value": 30},
            ], 10, inline=True, id="target-bf"),
        ]),
    ], style={"display": "grid", "gridTemplateColumns": "repeat(3, minmax(0, 1fr))", "gap": "24px"}),

    html.Div([
        dcc.Graph(id="power", style={"width": "50%", "minWidth": "0"}),
        dcc.Graph(id="bayes-design", style={"width": "50%", "minWidth": "0"}),
    ], style={"display": "flex"}),

    html.H3("Required sample size"),
    html.P(id="power-comment"),
    html.P(id="bayes-comment")

], style={"maxWidth": "1200px", "margin": "auto"})


@app.callback(
    Output("results", "children"),
    Output("distribution", "figure"),
    Output("power", "figure"),
    Output("bayes-design", "figure"),
    Output("power-comment", "children"),
    Output("bayes-comment", "children"),
    Input("n", "value"),
    Input("mean", "value"),
    Input("sd", "value"),
    Input("g", "value"),
    Input("desired-power", "value"),
    Input("alpha", "value"),
    Input("target-bf", "value")
)
def update(n, mean, sd, g, desired_probability, alpha, target_bf):
    """
    Update the simulation, design plots, and sample-size summaries.

    Parameters
    ----------
    n : int
        Participants per group.
    mean : float
        Control-group mean reaction time.
    sd : float
        Pooled reaction-time standard deviation.
    g : float
        Standardized effect estimated by Hedges' g.
    desired_probability : float
        Target power and BF design probability.
    alpha : float
        Frequentist significance level.
    target_bf : float
        BF10 evidence threshold.

    Returns
    -------
    tuple
        Result text, three figures, and two sample-size summaries.
    """
    df = simulate_stroop(n=n, ms_mean=mean, ms_sd=sd, g=g)

    t, p, _, _ = stroop_t_test(df, silent=True)
    bf10, interpretation = stroop_bayes_t_test(df, silent=True)
    current_power, required_n = stroop_power(n, g, desired_probability, alpha)
    current_bayes, required_bayes_n = stroop_bayes_design(n, g, desired_probability, target_bf)

    distribution = px.histogram(
        df,
        x="reaction_time_ms",
        color="group",
        barmode="overlay",
        opacity=0.6,
        nbins=30,
        title="Simulated reaction times",
        labels={"reaction_time_ms": "Reaction time (ms)", "group": "Group"},
    )

    ns = np.arange(10, 201)
    powers = [stroop_power(x, g, desired_probability, alpha)[0] for x in ns]
    bayes_probabilities = [stroop_bayes_probability(x, g, target_bf) for x in ns]

    power_plot = px.line(
        x=ns,
        y=powers,
        labels={
            "x": "Participants per group",
            "y": "Power"
        },
        title="Frequentist power",
    )

    power_plot.add_hline(y=desired_probability, line_dash="dash")
    power_plot.add_vline(x=n, line_dash="dash")
    power_plot.add_scatter(
        x=[n],
        y=[current_power],
        mode="markers",
        name="Current design"
    )

    bayes_plot = px.line(
        x=ns,
        y=bayes_probabilities,
        labels={"x": "Participants per group", "y": f"P(BF10 ≥ {target_bf})"},
        title="Bayes-factor design",
    )
    bayes_plot.add_hline(y=desired_probability, line_dash="dash")
    bayes_plot.add_vline(x=n, line_dash="dash")
    bayes_plot.add_scatter(x=[n], y=[current_bayes], mode="markers", name="Current design")

    if current_power >= desired_probability:
        power_comment = (
            f"Frequentist: power {current_power:.2f} ≥ {desired_probability:.2f}. "
            f"{required_n} participants per group are required."
        )
    else:
        missing = required_n - n
        power_comment = (
            f"Frequentist: power {current_power:.2f} < {desired_probability:.2f}. "
            f"{required_n} participants per group are required; add {missing} per group ({2 * missing} total)."
        )

    bayes_comment = (
        f"Bayesian: P(BF10 ≥ {target_bf}) = {current_bayes:.2f}. "
        f"{required_bayes_n} participants per group are required for probability {desired_probability:.2f}."
    )

    return (
        f"t = {t:.2f}, p = {p:.3f}; BF10 = {bf10:.2f}: {interpretation}.",
        distribution,
        power_plot,
        bayes_plot,
        power_comment,
        bayes_comment
    )


if __name__ == "__main__":
    app.run(debug=False)

File: src/r/power_analysis_app.R
invisible(lapply(c("ggplot2", "shiny"), function(package) {
  if (!requireNamespace(package, quietly = TRUE)) {
    install.packages(package, repos = "https://cloud.r-project.org")
  }
}))


if (!exists("simulate_stroop")) source(file.path("src", "r", "simulate.R"))
if (!exists("stroop_power")) source(file.path("src", "r", "analysis.R"))


ui <- shiny::fluidPage(
  shiny::tags$head(shiny::tags$style(shiny::HTML(
    ".container-fluid {max-width: 1200px; margin: auto;}"
  ))),
  shiny::titlePanel("Stroop simulation and design analysis"),
  shiny::p(
    paste(
      "Use the sample controls to simulate reaction times and inspect the group distributions.",
      "Then choose the design criteria to compare frequentist power with the probability",
      "of reaching a Bayes-factor threshold."
    )
  ),

  shiny::h3("Sample characteristics"),
  shiny::fluidRow(
    shiny::column(6, shiny::sliderInput("n", "Participants per group", 10, 200, 50, step = 10)),
    shiny::column(6, shiny::sliderInput("mean", "Control mean (ms)", 400, 800, 600, step = 25))
  ),
  shiny::fluidRow(
    shiny::column(6, shiny::sliderInput("sd", "Pooled SD (ms)", 25, 200, 100, step = 25)),
    shiny::column(6, shiny::sliderInput("g", "Hedges' g", 0.1, 1.5, HEDGES_G, step = 0.01))
  ),
  shiny::plotOutput("distribution", height = "400px"),
  shiny::textOutput("results"),

  shiny::h3("Design criteria"),
  shiny::fluidRow(
    shiny::column(4, shiny::sliderInput("desired", "Target probability", 0.6, 0.95, 0.8, step = 0.05)),
    shiny::column(4, shiny::sliderInput("alpha", "Frequentist alpha", 0.01, 0.10, 0.05, step = 0.01)),
    shiny::column(4, shiny::radioButtons(
      "target_bf", "Bayes factor threshold",
      choices = c("BF10 >= 3" = 3, "BF10 >= 10" = 10, "BF10 >= 30" = 30),
      selected = 10, inline = TRUE
    ))
  ),
  shiny::fluidRow(
    shiny::column(6, shiny::plotOutput("power")),
    shiny::column(6, shiny::plotOutput("bayes_design"))
  ),

  shiny::h3("Required sample size"),
  shiny::textOutput("power_comment"),
  shiny::textOutput("bayes_comment")
)


server <- function(input, output, session) {
  simulated <- shiny::reactive({
    simulate_stroop(input$n, ms_mean = input$mean, ms_sd = input$sd, g = input$g)
  })

  tests <- shiny::reactive({
    list(
      frequentist = stroop_t_test(simulated(), silent = TRUE),
      bayesian = stroop_bayes_t_test(simulated(), silent = TRUE)
    )
  })

  designs <- shiny::reactive({
    target_bf <- as.numeric(input$target_bf)
    list(
      frequentist = stroop_power(input$n, input$g, input$desired, input$alpha),
      bayesian = stroop_bayes_design(input$n, input$g, input$desired, target_bf),
      target_bf = target_bf
    )
  })

  curves <- shiny::reactive({
    ns <- 10:200
    target_bf <- as.numeric(input$target_bf)
    data.frame(
      n = ns,
      power = vapply(ns, .t_test_power, numeric(1), g = input$g, alpha = input$alpha),
      bayes = vapply(ns, stroop_bayes_probability, numeric(1), g = input$g, target_bf = target_bf)
    )
  })

  output$distribution <- shiny::renderPlot(stroop_histplot(simulated()))

  output$results <- shiny::renderText({
    result <- tests()
    sprintf(
      "t = %.2f, p = %.3f; BF10 = %.2f: %s.",
      result$frequentist$statistic, result$frequentist$p_value,
      result$bayesian$bf10, result$bayesian$interpretation
    )
  })

  output$power <- shiny::renderPlot({
    ggplot2::ggplot(curves(), ggplot2::aes(n, power)) +
      ggplot2::geom_line() +
      ggplot2::geom_hline(yintercept = input$desired, linetype = "dashed") +
      ggplot2::geom_vline(xintercept = input$n, linetype = "dashed") +
      ggplot2::geom_point(
        data = data.frame(n = input$n, power = designs()$frequentist$power), size = 2
      ) +
      ggplot2::coord_cartesian(ylim = c(0, 1)) +
      ggplot2::labs(title = "Frequentist power", x = "Participants per group", y = "Power") +
      ggplot2::theme_minimal()
  })

  output$bayes_design <- shiny::renderPlot({
    ggplot2::ggplot(curves(), ggplot2::aes(n, bayes)) +
      ggplot2::geom_line() +
      ggplot2::geom_hline(yintercept = input$desired, linetype = "dashed") +
      ggplot2::geom_vline(xintercept = input$n, linetype = "dashed") +
      ggplot2::geom_point(
        data = data.frame(n = input$n, bayes = designs()$bayesian$probability), size = 2
      ) +
      ggplot2::coord_cartesian(ylim = c(0, 1)) +
      ggplot2::labs(
        title = "Bayes-factor design", x = "Participants per group",
        y = sprintf("P(BF10 >= %s)", designs()$target_bf)
      ) +
      ggplot2::theme_minimal()
  })

  output$power_comment <- shiny::renderText({
    result <- designs()$frequentist
    if (result$power >= input$desired) {
      sprintf(
        "Frequentist: power %.2f >= %.2f. %d participants per group are required.",
        result$power, input$desired, result$required_n
      )
    } else {
      missing <- result$required_n - input$n
      sprintf(
        paste0(
          "Frequentist: power %.2f < %.2f. %d participants per group are required; ",
          "add %d per group (%d total)."
        ),
        result$power, input$desired, result$required_n, missing, 2 * missing
      )
    }
  })

  output$bayes_comment <- shiny::renderText({
    result <- designs()$bayesian
    sprintf(
      "Bayesian: P(BF10 >= %s) = %.2f. %d participants per group are required for probability %.2f.",
      designs()$target_bf, result$probability, result$required_n, input$desired
    )
  })
}


power_analysis_app <- shiny::shinyApp(ui, server)


if (sys.nframe() == 0L) shiny::runApp(power_analysis_app)
